platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/InstanceAndPublicType.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.declarativedsl
+
+import kotlin.reflect.KClass
+
+data class InstanceAndPublicType(val instance: Any?, val publicType: KClass<*>?) {
+    companion object {
+        val NULL = of(null, null)
+        val UNIT = of(Unit, Unit::class)
+
+        fun of(instance: Any?, publicType: KClass<*>?) = InstanceAndPublicType(instance, publicType)
+
+        fun unknownPublicType(instance: Any) = InstanceAndPublicType(instance, instance::class)
+    }
+
+    fun validate(errorMessageOnInstanceNull: () -> String): Pair<Any, KClass<*>> {
```

Comment:
❓ Why not return an `InstanceAndPublicType` instead of the `Pair` here? I think it would make call sites clearer

Reply:
The idea was that validation returns kind of an `InstanceAndPublicType`, but with non-null `instance` and `publicType` fields and I didn't want to create yet another type for that... Maybe there is a better solution for this, but the reasoning was this.

Reply:
Yep, the only way would be with two data classes like `NullableInstanceAndPublicType` and `InstanceAndPublicType`. In general it would be better but I see that there is currently a single usage site. You can leave it like that.